### DISCUSSION TITLE
How do we ensure that the filename externalized property is filled in before the CSVWriter and FFMPEGWriter nodes are run?

### DISCUSSION THREAD

**Discussion by kaelabruin** [bonsai-rx/bonsai#2034] [Q&A] [answered]
Hi!

I'm having some issues with my bonsai workflow. Here is the set up:

1. CSVWriter and FFMPEGWriter both need an initial filename in order for the workflow to run
2. We're updating the filename property to use the current date and timestamp via an externalized property
3. After running the workflow, the filename property on the CSVWriter and FFMPEGWriter nodes are overwritten with the formatted values as expected
4. The issue we're having is that the externalized property for the filename is not being evaluated before the CSVWriter and FFMPEGWriter nodes consume it, so it always uses the cached value from the last run

How do we ensure that the filename externalized property is filled in before the CSVWriter and FFMPEGWriter nodes are run?

Graphically: the highlighted nodes below need to run before all of the other nodes

![image](https://github.com/user-attachments/assets/eeef80d8-adea-487d-8ff3-9631769f65a0)

Thanks!

Kai

**Comment by bruno-f-cruz** [comment, marked as answer]
Avoid the branching pattern when initializing property values as stated in the docs. https://bonsai-rx.org/docs/articles/workflow-guidelines.html#property-initialization

Instead initialize the value to a subject and subscribe to it twice.

See an example here https://github.com/bonsai-rx/bonsai/discussions/1367#discussioncomment-5940478

**Reply by kaelabruin** [reply]
Hi Bruno, 

Thank you for the help, but I am honestly still a little confused as to how to apply this to my workflow. Would you mind editing this workflow to demonstrate?

[autostart_dates.n_cams=5.cams=1-5.crf=28.fps=50.v01.zip](https://github.com/user-attachments/files/17320224/autostart_dates.n_cams.5.cams.1-5.crf.28.fps.50.v01.zip)

Thanks!

Kai

**Reply by bruno-f-cruz** [reply]
I cant compile your workflow without your custom extensions code. But here's the gist:

![image](https://github.com/user-attachments/assets/76bc9b88-ccfe-4a0f-a648-aae6f31ae817)

**Reply by kaelabruin** [reply]
Thank you so much. Ok, the last thing I am wondering is what type / node in the toolbox are the timestamp values (both purple and green)? Is it simply workflow input and out?

**Reply by bruno-f-cruz** [reply]
They are BehaviorSubject and its respective reference. You can read more about them [here](https://bonsai-rx.org/docs/articles/subjects.html).

**Reply by kaelabruin** [reply]
Got it, thanks so much!
